File: moqt/internal/message/announce_ok_test.go
package message_test

import (
	"bytes"
	"testing"

	"github.com/qumo-dev/gomoqt/moqt/internal/message"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestAnnounceOkMessage_EncodeDecode(t *testing.T) {
	tests := map[string]struct {
		input message.AnnounceOkMessage
	}{
		"zero values": {
			input: message.AnnounceOkMessage{},
		},
		"non-relay endpoint": {
			input: message.AnnounceOkMessage{HopID: 0, ActiveCount: 3},
		},
		"relay with hop id": {
			input: message.AnnounceOkMessage{HopID: 0x3FFFFFFF, ActiveCount: 128},
		},
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			var buf bytes.Buffer

			require.NoError(t, tt.input.Encode(&buf))

			var decoded message.AnnounceOkMessage
			require.NoError(t, decoded.Decode(&buf))

			assert.Equal(t, tt.input, decoded)
		})
	}
}

func TestAnnounceOkMessage_DecodeErrors(t *testing.T) {
	tests := map[string]struct {
		data []byte
	}{
		"empty input":      {data: []byte{}},
		"missing fields":   {data: []byte{0x01, 0x00}}, // length 1, only HopID
		"trailing garbage": {data: []byte{0x03, 0x00, 0x00, 0xFF}},
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			var decoded message.AnnounceOkMessage
			err := decoded.Decode(bytes.NewReader(tt.data))
			assert.Error(t, err)
		})
	}
}

File: moq-web/src/internal/message/announce_ok_test.ts
import { assertEquals } from "@std/assert";
import { AnnounceOkMessage } from "./announce_ok.ts";
import { Buffer } from "@okdaichi/golikejs/bytes";

Deno.test("AnnounceOkMessage - encode/decode roundtrip", async (t) => {
	const testCases = {
		"zero values": { hopID: 0, activeCount: 0 },
		"non-relay with initial set": { hopID: 0, activeCount: 3 },
		"relay with hop id": { hopID: 12345, activeCount: 128 },
	};

	for (const [caseName, input] of Object.entries(testCases)) {
		await t.step(caseName, async () => {
			const buffer = Buffer.make(64);
			const message = new AnnounceOkMessage(input);
			assertEquals(await message.encode(buffer), undefined);

			const decoded = new AnnounceOkMessage({});
			assertEquals(await decoded.decode(buffer), undefined);
			assertEquals(decoded.hopID, input.hopID);
			assertEquals(decoded.activeCount, input.activeCount);
		});
	}

	await t.step("decode returns error on empty input", async () => {
		const buffer = Buffer.make(0);
		const decoded = new AnnounceOkMessage({});
		const err = await decoded.decode(buffer);
		assertEquals(err !== undefined, true);
	});
});
